theano/tensor/signal/pool.py

```diff
@@ -357,279 +415,368 @@ def c_headers(self):
     def c_code(self, node, name, inp, out, sub):
         if self.mode not in ('max', 'sum', 'average_exc_pad', 'average_inc_pad'):
             raise theano.gof.utils.MethodNotDefined()
+        nd = len(self.ds)
         x, = inp
         z, = out
+        total_ndim = node.inputs[0].ndim
+        non_pool_ndim = total_ndim - nd
         fail = sub['fail']
         ignore_border = int(self.ignore_border)
-        ds0, ds1 = self.ds
-        st0, st1 = self.st
-        pd0, pd1 = self.padding
+        ds = ','.join(map(str, self.ds))
+        st = ','.join(map(str, self.st))
+        pd = ','.join(map(str, self.padding))
         if self.openmp:
-            omp_parallel = '#pragma omp parallel for private(r_st, r_end, c_st, c_end, collector) schedule(static)'
+            # run in parallel over each pooling block
+            omp_parallel = '#pragma omp parallel for private(r_st, r_end, r_idx, i_idx, o_idx, collector) schedule(static)'
         else:
             omp_parallel = ''
         ccode = """
         int typenum = PyArray_ObjectType((PyObject*)%(x)s, 0);
-        int z_r, z_c; // shape of the output
-        int r, c; // shape of the padded_input
-        if(PyArray_NDIM(%(x)s)!=4)
+        int z[%(nd)s]; // shape of the output
+        int r[%(nd)s]; // shape of the padded_input
+        int ds[%(nd)s] = {%(ds)s};
+        int st[%(nd)s] = {%(st)s};
+        int pd[%(nd)s] = {%(pd)s};
+        if(PyArray_NDIM(%(x)s)!=%(total_ndim)s)
         {
-            PyErr_SetString(PyExc_ValueError, "x must be a 4d ndarray");
+            PyErr_SetString(PyExc_ValueError, "x must be a %(total_ndim)sD ndarray");
             %(fail)s;
         }
-        r = PyArray_DIMS(%(x)s)[2];
-        c = PyArray_DIMS(%(x)s)[3];
-        r += %(pd0)s * 2;
-        c += %(pd1)s * 2;
-        if (%(pd0)s != 0 && %(pd1)s != 0 && !%(ignore_border)s)
-            {
-              PyErr_SetString(PyExc_ValueError,
-                "padding must be (0,0) when ignore border is False");
-              %(fail)s;
-            }
+        for (int i=0; i<%(nd)s; i++)
+        {
+            r[i] = PyArray_DIMS(%(x)s)[%(non_pool_ndim)s + i] + 2 * pd[i];
+        }
+        """
+
+        if not self.ignore_border and any(self.padding[i] > 0 for i in xrange(nd)):
```

Comment:
There is a PR that make the padding as an input of the node. To help the conflict resolution of both PR, can you keep that in the C code?